### PR TITLE
Add AWS Terraform Quickstarter

### PR DESCRIPTION
**Pull request by nichtraunzer**
Here is the initial PR for the Terraform Jenkins agent and the AWS Terraform quickstarter.
Things we need to discuss: 
- name of quickstarter : inf-terraform-aws 
- name of agent : terraform (it will also support AZURE in the future)
- proper documentation

### CODE REVIEW

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:72]: Hardcoded stuff should be removed. `NO_PROXY` should be baked into the image using cluster-wide config (this is done automatically).

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:114]: Should probably not be logged?

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:140]: gitlab?

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:184]: same as above

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:228]: couldn't we use `kubernetes.io/basic-auth` to store both `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` in one secret.

File: inf-terraform-aws/README.md
Comment (michaelsauter) [inline inf-terraform-aws/README.md:1]: Naming?

File: inf-terraform-aws/files/AUTHORS
Comment (michaelsauter) [inline inf-terraform-aws/files/AUTHORS:1]: Not sure if this makes sense here, but it would be good to add this to the `OWNERS` filer

File: inf-terraform-aws/files/CONTRIBUTING.md
Comment (michaelsauter) [inline inf-terraform-aws/files/CONTRIBUTING.md:47]: URL?

File: inf-terraform-aws/files/CONTRIBUTING.md
Comment (michaelsauter) [inline inf-terraform-aws/files/CONTRIBUTING.md:65]: URL?

File: inf-terraform-aws/files/Makefile
Comment (michaelsauter) [inline inf-terraform-aws/files/Makefile:10]: ?

**Review by michaelsauter — COMMENTED** [review]
Just browsed this quickly. Generally looks like it's going in the right direction.

Seems there are quite a few hardcoded things - I stopped commenting everything there but please search for those hardcoded identifiers ...

File: inf-terraform-aws/files/CONTRIBUTING.md
Comment (metmajer) [reply inf-terraform-aws/files/CONTRIBUTING.md:47]: Let's not add instructions on how to contribute Blueprints yet. They are not part of ODS yet.

File: inf-terraform-aws/files/CONTRIBUTING.md
Comment (metmajer) [reply inf-terraform-aws/files/CONTRIBUTING.md:65]: See my comment regarding contribution guidelines.

File: inf-terraform-aws/files/Makefile
Comment (metmajer) [reply inf-terraform-aws/files/Makefile:10]: The quickstarter replaces the need for an upstream stack framework, since it is integrated into the quickstarter, which is now the source of truth. Release emails should be empty initially and documented as convenience in README.md. Please make sure that the user experience does not break if emails are undefined.

**Comment by metmajer** [comment]
Quick exchange of thoughts please @nichtraunzer @michaelsauter: should we name the agent IaaC so it can hold Ansible as well later on?

**Comment by michaelsauter** [comment]
What would be the overlap between this quickstarter and an ansible one? I do not know enough about this. In my view, the overlap determines if we should make it more general. Without knowing too much, I would lean towards keeping this named `terraform`.

**Comment by metmajer** [comment]
@michaelsauter the only reason to consolidate would be if we are comfortable with throwing Ansible and Terraform into a bucket IaaC. Otherwise, they have little overlaps.

File: common/jenkins-agents/terraform/ocp-config/bc.yml
Comment (nichtraunzer) [reply common/jenkins-agents/terraform/ocp-config/bc.yml]: copy/paste error - there are empty triggers in other agents too - we copied from golang agent

File: common/jenkins-agents/terraform/README.md
Comment (nichtraunzer) [reply common/jenkins-agents/terraform/README.md]: resolved https://github.com/nichtraunzer/ods-quickstarters/commit/0b904aba383c72ec5ab584146288a5cc85e943a1

File: common/jenkins-agents/terraform/README.md
Comment (nichtraunzer) [reply common/jenkins-agents/terraform/README.md]: https://github.com/nichtraunzer/ods-quickstarters/commit/0b904aba383c72ec5ab584146288a5cc85e943a1

File: common/jenkins-agents/terraform/ocp-config/bc.yml
Comment (nichtraunzer) [reply common/jenkins-agents/terraform/ocp-config/bc.yml]: https://github.com/nichtraunzer/ods-quickstarters/commit/0b904aba383c72ec5ab584146288a5cc85e943a1

File: inf-terraform-aws/README.md
Comment (nichtraunzer) [reply inf-terraform-aws/README.md]: https://github.com/nichtraunzer/ods-quickstarters/commit/0b904aba383c72ec5ab584146288a5cc85e943a1

File: inf-terraform-aws/files/AUTHORS
Comment (nichtraunzer) [reply inf-terraform-aws/files/AUTHORS]: https://github.com/nichtraunzer/ods-quickstarters/commit/0b904aba383c72ec5ab584146288a5cc85e943a1

**Comment by michaelsauter** [comment]
@metmajer Given this I'd vote for keeping this focused on Terraform.

**Comment by nichtraunzer** [comment]
> @michaelsauter the only reason to go with a single agent containing both would be if we are comfortable with throwing Ansible and Terraform into a bucket IaaC. Otherwise, they have little overlaps.

Basically I am open to that - but what about new releases of Ansible and/or Terraform - this would mean an high number of updates on the agent

**Comment by clemensutschig** [comment]
@nichtraunzer  - very nice .. still a lot of things to cleanup ... but really nice :)

**Comment by metmajer** [comment]
Thanks @michaelsauter for the good feedback above. We'll have all this cleaned up.

File: common/jenkins-agents/terraform/ocp-config/bc.yml
Comment (tbugfinder) [reply common/jenkins-agents/terraform/ocp-config/bc.yml]: https://github.com/opendevstack/ods-quickstarters/pull/515/commits/921ae1e534ffb4cf72b78f0f51e60e3473037d17
done.

File: common/jenkins-agents/terraform/ocp-config/is.yml
Comment (tbugfinder) [reply common/jenkins-agents/terraform/ocp-config/is.yml:14]: https://github.com/opendevstack/ods-quickstarters/pull/515/commits/921ae1e534ffb4cf72b78f0f51e60e3473037d17
done

Comment (tbugfinder) [reply common/jenkins-agents/terraform/ocp-config/is.yml:14]: @nichtraunzer - the branch should maybe resynced with master.

Comment (nichtraunzer) [reply common/jenkins-agents/terraform/ocp-config/is.yml:14]: done

File: inf-terraform-aws/files/CONTRIBUTING.md
Comment (nichtraunzer) [reply inf-terraform-aws/files/CONTRIBUTING.md]: removed

File: inf-terraform-aws/files/Makefile
Comment (nichtraunzer) [reply inf-terraform-aws/files/Makefile]: removed

File: inf-terraform-aws/files/Makefile
Comment (michaelsauter) [inline inf-terraform-aws/files/Makefile:13]: `.PHONY`?

File: inf-terraform-aws/files/Makefile
Comment (michaelsauter) [inline inf-terraform-aws/files/Makefile:15]: `.PHONY`?

File: inf-terraform-aws/files/VERSION
Comment (michaelsauter) [inline inf-terraform-aws/files/VERSION:1]: Is this used by the quickstarter? If so, how does this correlate to version e.g. set in Jira?

File: inf-terraform-aws/ocp.env
Comment (michaelsauter) [inline inf-terraform-aws/ocp.env:4]: I think this file is not needed. It's used in https://github.com/opendevstack/ods-quickstarters/blob/2f954388379b770419e00e20b2e81521ddafa0ec/common/scripts/create-component.sh#L128, which is only used to create OpenShift components - but you don't have any for this quickstarter, right?

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (metmajer) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:93]: @tbugfinder document generation has been removed from here, so we don't need this, right?

**Comment by metmajer** [comment]
Boom! Great comments by @michaelsauter! @nichtraunzer @tbugfinder :)

**Comment by nichtraunzer** [comment]
> Boom! Great comments by @michaelsauter! @nichtraunzer @tbugfinder :)

I guess finishing testing will last longer now ...

**Comment by metmajer** [comment]
> I guess finishing testing will last longer now ...

@nichtraunzer just stop raunzing and get to it :)

**Comment by nichtraunzer** [comment]
> > I guess finishing testing will last longer now ...
> 
> @nichtraunzer just stop raunzing and get to it :)
raunzing would sound different : "oida ... i mog nimma"

File: inf-terraform-aws/files/Makefile
Comment (nichtraunzer) [reply inf-terraform-aws/files/Makefile:14]: 3fc73db

File: inf-terraform-aws/files/VERSION
Comment (nichtraunzer) [reply inf-terraform-aws/files/VERSION:1]: 3fc73db

File: inf-terraform-aws/Jenkinsfile.template
Comment (tbugfinder) [reply inf-terraform-aws/Jenkinsfile.template:138]: correct

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:45]: git >=v2 was added for development cycle.

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:93]: Will drop it.

File: common/jenkins-agents/terraform/docker/Gemfile
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Gemfile:8]: Other's are dependencies and contraints are set by the ones set.

File: common/jenkins-agents/terraform/docker/Gemfile
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Gemfile:19]: Constraint's are set by inspec.

File: inf-terraform-aws/files/Gemfile
Comment (tbugfinder) [reply inf-terraform-aws/files/Gemfile:19]: Version constraints are set by kitchen-terraform, inspec.

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:84]: removed the second installation.
As tfenv is provided by default users can pick preferred version anyway.

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:108]: unfortunately not.

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7]: fixed.

File: common/jenkins-agents/terraform/docker/Dockerfile.ubi8
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Dockerfile.ubi8:135]: files are in sync (as good as possible).

File: common/jenkins-agents/terraform/docker/Dockerfile.rhel7
Comment (michaelsauter) [reply common/jenkins-agents/terraform/docker/Dockerfile.rhel7:84]: Cool. Maybe adapt the comment `#Install tfenv and multiple terraform versions` as well so no-one is confused later on? :)

**Review by michaelsauter — COMMENTED** [review]
Cool!

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile.template:78]: Do we really want to read this from a file?

Comment (nichtraunzer) [reply inf-terraform-aws/Jenkinsfile.template:78]: what is your concern?  we only read a reference - the access key itself is stored in OC secrets .

Comment (michaelsauter) [reply inf-terraform-aws/Jenkinsfile.template:78]: Ah cool! I got confused and thought you read the actual secret from the file. All good then.

Maybe I got confused because you read key AND secret. In theory you should also be able to reference just one secret that contains key and secret ... e.g. the basic auth type. What do you think? Might be easier to manage from a user perspective.

File: inf-terraform-aws/Jenkinsfile
Comment (michaelsauter) [inline inf-terraform-aws/Jenkinsfile:13]: Do you really want this? I'd think no OpenShift resources are required ...

File: inf-terraform-aws/Jenkinsfile.template
Comment (michaelsauter) [reply inf-terraform-aws/Jenkinsfile.template:78]: On that note, I don't find automation / description right now how a user would setup these credentials. Do you have more info on that?

Comment (nichtraunzer) [reply inf-terraform-aws/Jenkinsfile.template:78]: good point  ... supposed to be a manual step - helper script creates secrets in OC, user has to adjust values manually

File: inf-terraform-aws/Jenkinsfile
Comment (nichtraunzer) [reply inf-terraform-aws/Jenkinsfile]: copy paste error I guess ... removed it 254fc6d

File: common/jenkins-agents/terraform/docker/Gemfile
Comment (tbugfinder) [reply common/jenkins-agents/terraform/docker/Gemfile:1]: Few weeks ago I tested such a mirror using `bundle config --local mirror.https://rubygems.org http://odsnexus/repository/path`. Prerequisite is a proper repo setup. While testing nexus was slower than "native internet speed".
Is there a preset variable for it available, should it be an additional ENV settings or just a configuration option in case defined within the ODS project? For this project it would make sense to also provide a mirror for packagecloud.io.

File: inf-terraform-aws/Jenkinsfile.template
Comment (tbugfinder) [reply inf-terraform-aws/Jenkinsfile.template:138]: @michaelsauter Is ssh-agent plugin available in ODS?
https://www.jenkins.io/doc/pipeline/steps/ssh-agent/

File: inf-terraform-aws/Jenkinsfile.template
Comment (metmajer) [reply inf-terraform-aws/Jenkinsfile.template:8]: @nichtraunzer can you look into this comment?

File: inf-terraform-aws/Jenkinsfile
Comment (metmajer) [reply inf-terraform-aws/Jenkinsfile:3]: @nichtraunzer can you look into this comment?

File: inf-terraform-aws/ocp.env
Comment (metmajer) [reply inf-terraform-aws/ocp.env:4]: @nichtraunzer can you look into this comment?

**Review by metmajer — CHANGES_REQUESTED** [review]

**Comment by tbugfinder** [comment]
anything left ?
Azure is waiting..... (tick, tick, tick).

File: inf-terraform-aws/Jenkinsfile.template
Comment (tbugfinder) [reply inf-terraform-aws/Jenkinsfile.template:130]: There are dependencies which require even more changes then.

Comment (tbugfinder) [reply inf-terraform-aws/Jenkinsfile.template:130]: Well, git can even convert URIs like git::ssh automatically into https URIs using .gitconfig settings like
```
[url "https://username:token@mybbbserver.com/scm"]
        insteadOf = ssh://git@mybbbserver.com:<sshport>
```

**Comment by metmajer** [comment]
As soon as tests are finished, let's merge this PR and then re-create to target 3.x

**Review by nichtraunzer — COMMENTED** [review]
> @tbugfinder @nichtraunzer As soon as tests are finished, let's merge this PR and then re-create to target 3.x

I will have a quick look into the SSH vs HTTPS topic together with @michaelsauter tomorrow. In case we identify a quick solution which does not interfere with  the ongoing development (blueprints etc) we would like to implement it in this PR which means a minor delay. 
Otherwise  we create a separate issue and fix it with another PR.

File: inf-terraform-aws/files/variables.tf
Comment (tbugfinder) [inline inf-terraform-aws/files/variables.tf:28]: hm, outdated.

File: inf-terraform-aws/files/main.tf
Comment (tbugfinder) [inline inf-terraform-aws/files/main.tf:6]: hm, outdated

File: inf-terraform-aws/files/main.tf
Comment (tbugfinder) [inline inf-terraform-aws/files/main.tf:17]: including tags would show best practice.

Comment (nichtraunzer) [reply inf-terraform-aws/files/main.tf:17]: if you find a bug you may fix it :)

File: inf-terraform-aws/Jenkinsfile.template
Comment (nichtraunzer) [reply inf-terraform-aws/Jenkinsfile.template:130]: We fixed the ssh topic. Required changes are here 9024a42
Successful tests on an ODSiaB have shown we don't require an insteadOf statement ... 
I will test this on a prod EDP on Monday, too

File: inf-terraform-aws/README.md
Comment (metmajer) [inline inf-terraform-aws/README.md:1]: Docker image?

File: inf-terraform-aws/files/metadata.yml
Comment (metmajer) [inline inf-terraform-aws/files/metadata.yml:6]: Please add the ods version, 4.0, in this case

File: inf-terraform-aws/files/variables.tf
Comment (nichtraunzer) [reply inf-terraform-aws/files/variables.tf]: we cant kick computer_system_name - shouldnt it be part of the TIR

Comment (metmajer) [reply inf-terraform-aws/files/variables.tf]: yes, but the TIR does not get it from this parameter. you will need it, however, to label the cloud resources

File: inf-terraform-aws/files/main.tf
Comment (nichtraunzer) [reply inf-terraform-aws/files/main.tf:6]: please don't merge non working code !!!

https://github.com/nichtraunzer/ods-quickstarters/commit/207b0c29612ac818d995a91276c96e08112c70d9

**Review by metmajer — APPROVED** [review]
Cleaned up and tested. Awesome work @nichtraunzer @frankjoas @tbugfinder @michaelsauter 💯

**Pull request merged**
